### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by imeckler**
Creating on behalf of @Ltvvlz to have a PR for the initial implementation

### CODE REVIEW

File: marlin-succinct/circuits/src/compiled.rs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+*****************************************************************************************************************/
+
+use sprs::CsMat;
+use commitment::urs::URS;
+use algebra::{Field, PrimeField, PairingEngine};
+use ff_fft::{DensePolynomial, Evaluations, EvaluationDomain};
+pub use super::index::Index;
+
+pub struct Compiled<E: PairingEngine>
```
Comment (imeckler) [inline marlin-succinct/circuits/src/compiled.rs]: We will need the row/col/val polynomials evaluated on multiple domains. Check out the `Index` type [here](https://github.com/scipr-lab/marlin/blob/master/src/ahp/indexer.rs). I think it should be factored better into 3 records, one for each of A, B, C,  but I do believe we need all the values included in that type.

File: marlin-succinct/protocol/src/prover.rs [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+        // the first part of the argument needs the following oracles:
+        // alpha, betta[0] in order to compute RA and evaluate Z, ZA & ZB polynomials
+
+        let mut argument = RandomOracleArgument::<E::Fr>::new(index.oracle_params.clone());
+        
+        argument.commit_scalar(&E::Fr::one()); // this commit comes from the previous proof context
+        oracles.alpha = argument.challenge();
+        // commit the public input into the argument
+        argument.commit_slice(&witness.0[0..witness.1]);
+        oracles.betta[0] = argument.challenge();
```
Comment (imeckler) [inline marlin-succinct/protocol/src/prover.rs]: Just a note on spelling, should be `beta`.

File: marlin-succinct/protocol/src/prover.rs [hunk 3]
```diff
@@ -0,0 +1,599 @@
...
+
+        // evaluate Z, ZA & ZB polynomials
+        w_eval = w.evaluate(oracles.betta[0]);
+        za_eval = za.evaluate(oracles.betta[0]);
+        zb_eval = zb.evaluate(oracles.betta[0]);
+
+        // run the random oracle argument further to query
+        // oracles gamma[0..3]
+        argument.commit_scalar(&w_eval);
+        oracles.gamma[0] = argument.challenge();
```
Comment (imeckler) [inline marlin-succinct/protocol/src/prover.rs]: Can we use `eta_A, eta_B, eta_C`?

Comment (Ltvvlz) [reply marlin-succinct/protocol/src/prover.rs]: Sure. You prefer not to use oracle arrays like eta[0..3] but rather eta_A, eta_B, eta_C ?

File: marlin-succinct/protocol/src/prover.rs (hunk 2 above)
Comment (Ltvvlz) [reply marlin-succinct/protocol/src/prover.rs]: Thank you for spotting...

File: marlin-succinct/circuits/src/compiled.rs (hunk 1 above)
Comment (Ltvvlz) [reply marlin-succinct/circuits/src/compiled.rs]: OK, investigating this.

File: marlin-succinct/protocol/src/batch_verifier.rs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    (
+        &self,
+        index: &Index<E>,
+        rng: &mut dyn RngCore
+    ) -> Result<bool, ProofError>
+    {
+        let mut batch = vec![Vec::new(), Vec::new(), Vec::new()];
+        for proof in self.batch.iter()
+        {
+            match proof
```
Comment (imeckler) [inline marlin-succinct/protocol/src/batch_verifier.rs]: I think you can use the `?` operator to make this code less nested: https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/the-question-mark-operator-for-easier-error-handling.html

File: marlin-succinct/protocol/src/prover.rs [hunk 5]
```diff
@@ -0,0 +1,599 @@
...
+            }
+        ).collect();
+        r.reverse();
+        let ra = DensePolynomial::<E::Fr>::from_coefficients_vec(r);
+
+        // compute commitments, evaluations and openings for the first sumcheck argument
+        // proceed with the random oracle argument further to query additional oracles
+        // betta[1], batch[0]
+
+        match Self::sumcheck_1_compute (index, &ra, &zv, &z, &oracles)
```
Comment (imeckler) [inline marlin-succinct/protocol/src/prover.rs]: I believe you can similarly use the `?` operator here https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/the-question-mark-operator-for-easier-error-handling.html

File: marlin-succinct/protocol/src/prover.rs (hunk 3 above)
Comment (imeckler) [reply marlin-succinct/protocol/src/prover.rs]: Maybe a record with `a, b, c` fields? I would rather the not leave the mapping {0, 1, 2} <-> {A, B, C} implicit.

File: marlin-succinct/protocol/src/prover.rs [hunk 6]
```diff
@@ -134,29 +134,29 @@ impl<E: PairingEngine> ProverProof<E>
...
+        w_eval = w.evaluate(oracles.beta[0]);
+        za_eval = za.evaluate(oracles.beta[0]);
+        zb_eval = zb.evaluate(oracles.beta[0]);
 
         // run the random oracle argument further to query
-        // oracles gamma[0..3]
+        // oracles eta[0..3]
         argument.commit_scalar(&w_eval);
-        oracles.gamma[0] = argument.challenge();
+        oracles.eta[0] = argument.challenge();
```
Comment (imeckler) [inline marlin-succinct/protocol/src/prover.rs]: You can absorb the commitments to `w, z_A, z_B` before squeezing `alpha` and the `eta` values -- it's better for efficiency to not switch between absorbing and squeezing.

Comment (imeckler) [reply marlin-succinct/protocol/src/prover.rs]: Oh actually I don't quite understand what's going on, it seems things are not in the correct order here.

Comment (imeckler) [reply marlin-succinct/protocol/src/prover.rs]: I believe the order should be as in https://github.com/CodaProtocol/coda/blob/double-marlin-verifier/src/lib/double_marlin/double_marlin.ml#L202. And it is not necessary to absorb any polynomial evalutions

File: marlin-succinct/protocol/src/prover.rs [hunk 7]
```diff
@@ -172,189 +172,171 @@ impl<E: PairingEngine> ProverProof<E>
...
+        sigma2 = g2.coeffs[0];
+        g2.coeffs.remove(0);
+
+        // evaluate H2, G2 polynomials
+        h2_eval = h2.evaluate(oracles.beta[1]);
+        g2_eval = g2.evaluate(oracles.beta[1]);
+
+        // query batch[1], beta[2] random oracles from the argument context
+        argument.commit_scalar(&h2_eval);
+        oracles.batch[1] = argument.challenge();
```
Comment (imeckler) [inline marlin-succinct/protocol/src/prover.rs]: There is no need to have separate challenges for all the batches. You can just sample one at the end of the prover sending all the polynomial commitments and messages (I call it [xi here](https://github.com/CodaProtocol/coda/blob/double-marlin-verifier/src/lib/double_marlin/double_marlin.ml#L239))

File: marlin-succinct/protocol/src/prover.rs [hunk 8]
```diff
@@ -187,13 +187,13 @@ impl<E: PairingEngine> ProverProof<E>
             zb_comm,
             h1_comm,
             g1_comm,
-            h2_comm : index.urs.commit(&h2, index.h_group.size())?,
-            g2_comm : index.urs.commit(&g2, index.h_group.size()-1)?,
-            h3_comm : index.urs.commit(&h3, index.h_group.size()*6)?,
-            g3_comm : index.urs.commit(&g3, index.h_group.size()-1)?,
+            h2_comm: index.urs.commit(&h2, index.h_group.size())?,
+            g2_comm: index.urs.commit(&g2, index.h_group.size()-1)?,
+            h3_comm: index.urs.commit(&h3, index.h_group.size()*6)?,
```
Comment (imeckler) [inline marlin-succinct/protocol/src/prover.rs]: should be `index.k_group.size()` I think?

File: marlin-succinct/protocol/src/prover.rs [hunk 9]
```diff
@@ -187,13 +187,13 @@ impl<E: PairingEngine> ProverProof<E>
...
+            g3_comm: index.urs.commit(&g3, index.h_group.size()-1)?,
```
Comment (imeckler) [inline marlin-succinct/protocol/src/prover.rs]: same here

**Pull request merged**